addons/misra.py

```diff
@@ -2740,36 +2740,70 @@ def misra_12_3(self, data):
                     prev = prev.previous
 
     def misra_12_4(self, cfg):
-        for expr in cfg.tokenlist:
+        if not cfg.tokenlist:
+            return
+        expr = cfg.tokenlist[0]
+        while expr:
             if not expr.astOperand2 or not expr.astOperand1:
```

Comment:
I think you should break out the code to check a token into a helper function:

    def misra_12_4_check_expr(self, expr):
        if not expr.astOperand2 or not expr.astOperand1:
            return
        if expr.valueType is None:
            return
        ....
        if not is_constant_integer_expression(expr):
             return
        if expr.str == '+' and op1 + op2 > max_value:
            self.reportError(expr, 12, 4)
        elif expr.str == '-' and op1 - op2 < 0:
            self.reportError(expr, 12, 4)
        elif expr.str == '*' and op1 * op2 > max_value:
            self.reportError(expr, 12, 4)        


Then the while loop can be more like:

    while ..
         if expr.str == '?':
             .. handle ternary operator ..
         else:
             self.misra_12_4_check_expr(expr)
         

This makes it possible to write a `_recursive_check` that calls `misra_12_4_check_expr` in a tree recursively..